File: template-method-pattern/report-generator/report-generator.ts
interface Payload {
  [key: string]: number;
}

abstract class ReportGenerator {
  generateReport() {
    const data = this.fetchData();
    const formattedData = this.formatData(data);
    this.printReport(formattedData);
  }

  abstract fetchData(): Payload;

  abstract formatData(data: Payload): string;

  printReport(formattedData: string) {
    console.log("Report: ", formattedData);
  }
}

class SalesReport extends ReportGenerator {
  fetchData(): Payload {
    return { sales: 1200, profit: 800 };
  }

  formatData(data: Payload): string {
    return `Sales: ${data.sales}, Profit: ${data.profit}`;
  }
}

class UserActivityReport extends ReportGenerator {
  fetchData(): Payload {
    return { users: 200, active: 30 };
  }

  formatData(data: Payload): string {
    return `Total Users: ${data.users}, Active users: ${data.active}`;
  }
}

// Dry run
let report = new SalesReport();
report.generateReport();

report = new UserActivityReport();
report.generateReport();

File: template-method-pattern/report-generator/report-generator.rs
trait ReportGenerator {
    // Template method: defines the skeleton of the algorithm
    fn generate_report(&self) {
        let data = self.fetch_data();
        let formatted_data = self.format_data(&data);
        self.print_report(&formatted_data);
    }

    // Abstract method to fetch data
    fn fetch_data(&self) -> String;
    fn format_data(&self, data: &str) -> String;

    fn print_report(&self, formatted_data: &str) {
        println!("Report:\n{}", formatted_data);
    }
}

// Concrete implementations of ReportGenerator
struct SalesReport;

impl ReportGenerator for SalesReport {
    fn fetch_data(&self) -> String {
        "Sales data".to_string()
    }

    fn format_data(&self, data: &str) -> String {
        format!("Sales: 1000$, Profit: 200$, data = {}", data)
    }
}

struct UserActivityReport;

impl ReportGenerator for UserActivityReport {
    fn fetch_data(&self) -> String {
        "User activity data".to_string()
    }

    fn format_data(&self, data: &str) -> String {
        format!("Active users: 1000, Inactive users: 200, data = {}", data)
    }
}

fn main() {
    let sales_report = SalesReport;
    sales_report.generate_report();

    let user_activity_report = UserActivityReport;
    user_activity_report.generate_report();
}
